File: source/javascript/buffer_builder.js
// Growing buffer that doubles when full
// Write "Hello, World!" as bytes, then read back
let capacity = 4;
let size = 0;
let buf = new Uint8Array(capacity);

function bufGrow() {
    const newCap = capacity * 2;
    const newBuf = new Uint8Array(newCap);
    newBuf.set(buf.subarray(0, size));
    capacity = newCap;
    buf = newBuf;
}

function writeByte(b) {
    if (size >= capacity) bufGrow();
    buf[size++] = b;
}

function readBack() {
    return Buffer.from(buf.subarray(0, size)).toString('utf8');
}

const message = "Hello, World!";
const bytes = Buffer.from(message, 'utf8');
for (const b of bytes) writeByte(b);

console.log(`wrote ${size} bytes`);
console.log(`capacity: ${capacity}`);
const reconstructed = readBack();
console.log(`reconstructed: ${reconstructed}`);

File: source/rust/buffer_builder.rs
struct BufferBuilder {
    data: Vec<u8>,
    capacity: usize,
    size: usize,
}

impl BufferBuilder {
    fn new(capacity: usize) -> Self {
        BufferBuilder { data: vec![0u8; capacity], capacity, size: 0 }
    }

    fn write_byte(&mut self, b: u8) {
        if self.size >= self.capacity {
            let new_cap = self.capacity * 2;
            self.data.resize(new_cap, 0);
            self.capacity = new_cap;
        }
        self.data[self.size] = b;
        self.size += 1;
    }

    fn read_back(&self) -> String {
        String::from_utf8_lossy(&self.data[..self.size]).to_string()
    }
}

fn main() {
    let mut bb = BufferBuilder::new(4);
    let message = "Hello, World!";
    for &b in message.as_bytes() {
        bb.write_byte(b);
    }
    println!("wrote {} bytes", bb.size);
    println!("capacity: {}", bb.capacity);
    let reconstructed = bb.read_back();
    println!("reconstructed: {}", reconstructed);
}
